lib/spack/spack/solver/cache.py

```diff
@@ -0,0 +1,163 @@
+# Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+import hashlib
+import json
+import pathlib
+from contextlib import contextmanager
+from typing import List, Tuple, Union
+
+import llnl.util.tty as tty
+
+import spack
+import spack.config
+import spack.paths
+import spack.util.path
+from spack.solver.asp import Result
+
+
+class ConcretizationCache:
+    """Store for Spack concretization results and statistics
+
+    Serializes solver result objects and statistics to json and stores
+    at a given endpoint in a cache associated by the sha256 of the
+    asp problem and the involved control files.
+    """
+
+    def __init__(self, root: Union[str, None] = None):
+        if not root:
+            root = spack.config.get(
+                "config:concretization_cache", spack.paths.default_conc_cache_path
+            )
+        self.root = pathlib.Path(spack.util.path.canonicalize_path(root))
+        self.cleanup()
+
+    def cleanup(self):
+        # TODO: determine a better default
+        if self.root.exists():
+            entry_limit = spack.config.get("config:concretization_cache_limit", 1000)
+            entries = [x for x in self.cache_entries()]
```

Comment:
I think it's both brittle and very inefficient to traverse the file system tree here. I certainly don't want Spack to do another 1000 stat calls on top of the current 7K for package.py files.

Instead, keep track of the total number of files added and bytes written when `store` is called. That means you need a write lock for a file with totals.

Reply:
Inefficient for sure, I just wanted _something_ here as a cleanup operation so we could iterate on it in review. This is also terrible because it's run every time the conc cache is constructed, which should not go in as is, would love a suggestion for where/when we should be running this. 

Will refactor approach. There's some other metadata that would be a useful store, and I'd also like a pruning policy based on whether or not a given concretization is even possible anymore given changes to packages/lp files over time.